src/ui/VirtualFileRenderer/VirtualFileRenderer.spec.tsx

```diff
@@ -0,0 +1,361 @@
+    this.callback([{ target: { getAttribute: () => ({ scrollWidth: 100 }) } }])
+  }
+  unobserve() {
+    // do nothing
+  }
+  disconnect() {
+    // do nothing
+  }
+}
+global.window.ResizeObserver = ResizeObserver
+
+const code = `<Breadcrumb
+    paths={[
+    { pageName: 'owner', text: owner },
+    { pageName: 'repo', text: repo },
+    ...treePaths,
+    {..props}
+    ]}
+/>`
+
+const coverageData = {
+  0: 'H',
+  1: 'M',
+  2: 'P',
+  3: 'H',
+  4: 'M',
+  5: 'P',
+} as unknown as Dictionary<'H' | 'M' | 'P'>
+
+let testLocation: ReturnType<typeof useLocation>
+const wrapper =
+  (initialEntry = '/'): React.FC<React.PropsWithChildren> =>
+  ({ children }) => (
+    <MemoryRouter initialEntries={[initialEntry]}>
+      <Route path="/">{children}</Route>
+      <Route
+        path="*"
+        render={({ location }) => {
+          testLocation = location
+          return null
+        }}
+      />
+    </MemoryRouter>
+  )
+
+describe('VirtualFileRenderer', () => {
+  function setup() {
+    const user = userEvent.setup()
+
+    return { user }
+  }
+
+  it('renders the code', () => {
+    render(
+      <VirtualFileRenderer
+        code={code}
+        coverage={coverageData}
+        fileName="tsx"
+      />,
+      {
+        wrapper: wrapper(),
+      }
+    )
+
+    const codeBlock = screen.getByText(/<Breadcrumb/)
+    expect(codeBlock).toBeInTheDocument()
+  })
+
+  it('renders line numbers', () => {
+    render(
+      <VirtualFileRenderer
+        code={code}
+        coverage={coverageData}
+        fileName="tsx"
+      />,
+      {
+        wrapper: wrapper(),
+      }
+    )
+
+    const lineNumbers = screen.getAllByText(/\d+/)
+    expect(lineNumbers).toHaveLength(8)
+  })
+
+  describe('covered lines', () => {
+    it('uses the correct color', () => {
+      render(
+        <VirtualFileRenderer
+          code={code}
+          coverage={coverageData}
+          fileName="tsx"
+        />,
+        {
+          wrapper: wrapper(),
+        }
+      )
+
+      const coveredLine = screen.getByText(0)
+      expect(coveredLine).toBeInTheDocument()
+      expect(coveredLine).toHaveClass('bg-ds-coverage-covered')
+    })
+  })
+
+  describe('uncovered lines', () => {
+    it('uses the correct color', () => {
+      render(
+        <VirtualFileRenderer
+          code={code}
+          coverage={coverageData}
+          fileName="tsx"
+        />,
+        {
+          wrapper: wrapper(),
+        }
+      )
+
+      const uncoveredLine = screen.getByText(1)
+      expect(uncoveredLine).toBeInTheDocument()
+      expect(uncoveredLine).toHaveClass('bg-ds-coverage-uncovered')
+    })
+  })
+
+  describe('partial lines', () => {
+    it('uses the correct color', () => {
+      render(
+        <VirtualFileRenderer
+          code={code}
+          coverage={coverageData}
+          fileName="tsx"
+        />,
+        {
+          wrapper: wrapper(),
+        }
+      )
+
+      const partialLine = screen.getByText(2)
```

Comment:
Is there a better way you can query this that is more future proof? Right now if you were to change the test setup, this would break.

Reply:
Hrrm off the top of my head, I think the closet thing would be something along with these lines:

```
      const coveredLine = virtualOverlay.querySelector(
        '.bg-ds-coverage-covered'
      )
      expect(coveredLine).toHaveClass('bg-ds-coverage-covered')
```